keps/sig-apps/3939-allow-recreation-when-fully-terminated/README.md

```diff
@@ -0,0 +1,699 @@
+  - Components depending on the feature gate: kube-controller-manager
+
+###### Does enabling the feature change any default behavior?
+
+For enabling JobRecreatePodsWhenFailed:
+
+a) Count the number of terminating pods and populate in JobStatus
+b) With `RecreationPodsWhen: Failed` specified, pods will only be recreated when they are fully terminated.
+
+This could potentially make jobs (where pods are terminated) slower due to waiting for terminating pods
+to be fully deleted.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+Yes.
+<!--
+Describe the consequences on existing workloads (e.g., if this is a runtime
+feature, can it break the existing applications?).
+
+Feature gates are typically disabled by setting the flag to `false` and
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+Terminating pods will no longer be tracked so terminating pods will be considered deleted and new pods will be created.
+
+###### Are there any tests for feature enablement/disablement?
+
+Yes. Unit tests will include the fields off/on and verify behavior.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+#### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+#### What specific metrics should inform a rollback?
+
+<!--
+What signals should users be paying attention to when the feature is young
+that might indicate a serious problem?
+-->
+
+#### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+#### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+<!--
+Even if applying deprecation policies, they may still surprise some users.
+-->
+
+### Monitoring Requirements
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+-->
+
+#### How can an operator determine if the feature is in use by workloads?
+
+<!--
+Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
```

Comment:
I'd consider adding a metric or reusing one of the pre-existing ones, if possible.

Reply:
Will update next week.

Reply:
I'm not really sure how an operator can determine if this feature is in use by a metric.  I think we could add one for "average time to create pods once terminated started" but not sure if that is a good metric or not.

Reply:
We don't need a metric for every feature.
But it might be valuable to have a metric for number of pod creations, perhaps with a label about the operation that triggered it (new, recreateImmediate, recreateAfterTerminal)?